parametric_stellarator.py

```diff
@@ -380,45 +507,100 @@ def parametric_stellarator(
+                'facet_tol': maximum distance a facet may be from surface of
+                    CAD representation for Cubit export (float, defaults to
+                    None),
+                'len_tol': maximum length of facet edge for Cubit export
+                    (float, defaults to None),
+                'norm_tol': maximum change in angle between normal vector of
+                    adjacent facets (float, defaults to None),
+                'min_mesh_size': minimum mesh element size for Gmsh export
+                    (float, defaults to 5.0),
+                'max_mesh_size': maximum mesh element size for Gmsh export
+                    (float, defaults to 20.0),
+                'volume_atol': absolute volume tolerance to allow when matching
+                    parts in intermediate BREP file with CadQuery parts for
+                    Gmsh export(float, defaults to 0.00001),
+                'center_atol': absolute center coordinates tolerance to allow
+                    when matching parts in intermediate BREP file with CadQuery
+                    parts for Gmsh export (float, defaults to 0.00001),
+                'bounding_box_atol': absolute bounding box tolerance  to allow
+                    when matching parts in intermediate BREP file with CadQuery
+                    parts for Gmsh export (float, defaults to 0.00001).
+            }
+        magnets (dict): dictionary of magnet parameters including
+            {
+                'file': path to magnet coil point-locus data file (str),
+                'cross_section': coil cross-section definition (list),
+                'start': starting line index for data in file (int),
+                'stop': stopping line index for data in file (int),
+                'name': name to use for STEP export (str),
+                'h5m_tag': material tag to use in H5M neutronics model (str)
+            }
+            For the list defining the coil cross-section, the cross-section
+            shape must be either a circle or rectangle. For a circular
+            cross-section, the list format is
+            ['circle' (str), radius (float, cm)]
+            For a rectangular cross-section, the list format is
+            ['rectangle' (str), width (float, cm), thickness (float, cm)]
     """
+    # Signal new stellarator build
+    logger.info('New stellarator build')
+    
     # Check if total toroidal extent exceeds 360 degrees
     if gen_periods > num_periods:
+        logger.debug(
+            'Number of requested periods to generate exceeds number in '
+            'stellarator geometry')
         raise ValueError(
             'Number of requested periods to generate exceeds number in '
             'stellarator geometry'
         )
 
     # Check that h5m_export has an appropriate value
-    if h5m_export != None and h5m_export != 'Cubit' and h5m_export != 'Gmsh':
+    if export['h5m_export'] not in [None, 'Cubit', 'Gmsh']:
+        logger.debug(
```

Comment:
What I ended up doing instead is something of the form
```
# Check that h5m_export has an appropriate value
    if export['h5m_export'] not in [None, 'Cubit', 'Gmsh']:
        try:
            raise ValueError(
                'h5m_export must be None or have a string value of \'Cubit\' '
                'or \'Gmsh\''
            )
        except:
            logger.error('Exception occured', exc_info = True)
            sys.exit()
```
Do you think this works well?

Reply:
I think that's always going to result in calling `logger.error`. Raising an exception inside the `try:` block with a naked `except:` will capture all raise exceptions/errors (including the one in the `try:` block) and execute the code in the `except:` block.

What I was thinking is more like an additional method that you can throw an error message at and it will either log the message, raise an exception, or both as needed.

```python
def error(self, message, exception=None):
    # might want to get a default logger here instead as discussed in the software meeting
    if self.logger:
        logger.error(message)
        sys.exit()
    if exception is not None:
        raise exception(message)
```

Or something close to that. Just something that keeps you from having to create the message more than once since

Reply:
Yeah! My bad that we didn't discuss this today. It's slightly related to the `logger` discussion.  I think your code should only raise exceptions.  The code that calls it should decide how to handle this exceptions, either:
• ignore it so that the exception continues up the stack
• catch it and put the message onto the log
• catch it, put the message on the log and continue to raise it

Logging + raising an exception at the same time seems to be defeating the purpose of exceptions.

Reply:
My goal was to have logger.error get called every time. I formatted it this way so that when the error occurs, instead of just raising a ValueError, the logger would also capture it along with the traceback, and from what I found online, this was the only way to do so.

Reply:
I think this gets into the philosophy of exceptions a little.  Fundamentally, exceptions are a way of raising an error/warning state up the call stack to be handled by the code that called it.  Because the philosophy is for the calling code to handle the exception, it should be that calling code that decides whether or not the message ends up in the log.  Forcing every exception message to also end up in the log kind of undermines that philosophy because it means that the log will get that message regardless of how the calling code wants to handle the exception.

Reply:
Based on an internet perusal, it seems like assert() would fit nicely here. Although this is different from Patrick's suggestion, how do we feel about handling the exceptions like
```
try:
    assert(
        export['h5m_export'] not in [None, 'Cubit', 'Gmsh']
    )
except AssertionError as err:
    logger.exception(
        'h5m_export must be None or have a string value of \'Cubit\' or \'Gmsh\''
    )
    raise err
```

Reply:
Good point, @gonuke. Probably don't want all of these exception messages polluting the log if they are ignored further up the stack. I'm not sure how you would capture the stack output in the logger only if the exception is raised. Never attempted that before, but maybe we could look into it at the next software meeting.

@connoramoreno I understand you want to get the stack trace to the log here, but this approach seems a little convoluted. Any reason not to use the `traceback` module? Is there other info you're looking to capture in the log?

Reply:
>  Mainly, I just figured that if we're going to have a log, it could be useful to capture the specific error and traceback. 

I think it's a good idea, but it might depend on your definition of a log and what its purpose is. Goes back to our discussion on verbosity levels maybe.

Reply:
I'd like to resolve this sooner rather than later but maybe it would be best to discuss this at the software meeting

Reply:
I definitely don't want to slow progress here, so don't consider any of my comments to be blocking plz. I'll leave the decision on logging & exceptions to you and @gonuke. Hopefully the `traceback` module is helpful in simplifying things for you!

Reply:
Using your custom `logger` setup, this StackOverflow answer seemed to work: https://stackoverflow.com/a/37992988

```
try:
    <something that doesn't work>
except:
    logger.error("h5m_export must be None, 'Cubit', or 'Cmsh'", exc_info=e)
```

This outputs the error and traceback to both the screen and the logfile.